---
title: Avsluta fel | Microsoft Docs
ms.custom: ''
ms.date: 09/13/2016
ms.reviewer: ''
ms.suite: ''
ms.tgt_pltfrm: ''
ms.topic: article
ms.assetid: b804e738-aefa-41bb-9649-f9ed897fd96c
caps.latest.revision: 8
ms.openlocfilehash: d1967fe7996f75ec5229920f7ec49aa5ff6bdbfd
ms.sourcegitcommit: debd2b38fb8070a7357bf1a4bf9cc736f3702f31
ms.translationtype: MT
ms.contentlocale: sv-SE
ms.lasthandoff: 12/05/2019
ms.locfileid: "72359206"
---
# <a name="terminating-errors"></a><span data-ttu-id="b49de-102">Fel som avbryter körningen</span><span class="sxs-lookup"><span data-stu-id="b49de-102">Terminating Errors</span></span>

<span data-ttu-id="b49de-103">I det här avsnittet beskrivs den metod som används för att rapportera avslutande fel.</span><span class="sxs-lookup"><span data-stu-id="b49de-103">This topic discusses the method used to report terminating errors.</span></span> <span data-ttu-id="b49de-104">Den beskriver också hur du anropar-metoden från cmdleten och diskuterar undantagen som kan returneras av Windows PowerShell-körningsmiljön när metoden anropas.</span><span class="sxs-lookup"><span data-stu-id="b49de-104">It also discusses how to call the method from within the cmdlet, and it discusses the exceptions that can be returned by the Windows PowerShell runtime when the method is called.</span></span>

<span data-ttu-id="b49de-105">När ett avbrotts fel inträffar ska cmdleten rapportera felet genom att anropa metoden [system. Management. Automation. cmdlet. Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) .</span><span class="sxs-lookup"><span data-stu-id="b49de-105">When a terminating error occurs, the cmdlet should report the error by calling the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method.</span></span> <span data-ttu-id="b49de-106">Den här metoden gör att cmdleten kan skicka en felpost som beskriver det villkor som orsakade det avslutande felet.</span><span class="sxs-lookup"><span data-stu-id="b49de-106">This method allows the cmdlet to send an error record that describes the condition that caused the terminating error.</span></span> <span data-ttu-id="b49de-107">Mer information om fel poster finns i [fel poster för Windows PowerShell](./windows-powershell-error-records.md).</span><span class="sxs-lookup"><span data-stu-id="b49de-107">For more information about error records, see [Windows PowerShell Error Records](./windows-powershell-error-records.md).</span></span>

<span data-ttu-id="b49de-108">När metoden [system. Management. Automation. cmdlet. Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) anropas, stoppas körningen av pipelinen permanent i Windows PowerShell-körningsmiljön och ett [system. Management. Automation. PipelineStoppedException](/dotnet/api/System.Management.Automation.PipelineStoppedException) -undantag genereras.</span><span class="sxs-lookup"><span data-stu-id="b49de-108">When the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method is called, the  Windows PowerShell runtime permanently stops the execution of the pipeline and throws a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span> <span data-ttu-id="b49de-109">Eventuella efterföljande försök att anropa [system. Management. Automation. cmdlet. WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [system. Management. Automation. cmdlet. WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)eller flera andra API: er medför att dessa anrop genererar ett [system. Management. Automation. PipelineStoppedException](/dotnet/api/System.Management.Automation.PipelineStoppedException) -undantag.</span><span class="sxs-lookup"><span data-stu-id="b49de-109">Any subsequent attempts to call [System.Management.Automation.Cmdlet.WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError), or several other APIs causes those calls to throw a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span>

<span data-ttu-id="b49de-110">Undantags felet [system. Management. Automation. PipelineStoppedException](/dotnet/api/System.Management.Automation.PipelineStoppedException) kan också inträffa om en annan cmdlet i pipeline rapporterar ett avslutande fel, om användaren har bett att stoppa pipelinen eller om pipelinen har stoppats innan den slutförts av någon anledning.</span><span class="sxs-lookup"><span data-stu-id="b49de-110">The [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception can also occur if another cmdlet in the pipeline reports a terminating error, if the user has asked to stop the pipeline, or if the pipeline has been halted before completion for any reason.</span></span> <span data-ttu-id="b49de-111">Cmdleten behöver inte fånga upp undantaget [system. Management. Automation. PipelineStoppedException](/dotnet/api/System.Management.Automation.PipelineStoppedException) , om det inte måste rensas öppna resurser eller det interna läget.</span><span class="sxs-lookup"><span data-stu-id="b49de-111">The cmdlet does not need to catch the [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception unless it must clean up open resources or its internal state.</span></span>

<span data-ttu-id="b49de-112">Cmdletar kan skriva valfritt antal utgående objekt eller icke-avslutande fel innan ett avslutande fel rapporteras.</span><span class="sxs-lookup"><span data-stu-id="b49de-112">Cmdlets can write any number of output objects or non-terminating errors before reporting a terminating error.</span></span> <span data-ttu-id="b49de-113">Det avslutande felet stoppar dock pipelinen permanent och inga ytterligare utdata, avslutande fel eller icke-avslutande fel kan rapporteras.</span><span class="sxs-lookup"><span data-stu-id="b49de-113">However, the terminating error permanently stops the pipeline, and no further output, terminating errors, or non-terminating errors can be reported.</span></span>

<span data-ttu-id="b49de-114">-Cmdletar kan anropa [system. Management. Automation. cmdlet. Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) endast från den tråd som anropade metoden [system. Management. Automation. cmdlet. BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [system. Management. Automation. cmdlet. ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)eller [system. Management. Automation. cmdlet. EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) för bearbetnings metoden för indata.</span><span class="sxs-lookup"><span data-stu-id="b49de-114">Cmdlets can call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) only from the thread that called the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) input processing method.</span></span> <span data-ttu-id="b49de-115">Försök inte anropa [system. Management. Automation. cmdlet. Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) eller [system. Management. Automation. cmdlet. WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) från en annan tråd.</span><span class="sxs-lookup"><span data-stu-id="b49de-115">Do not attempt to call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) or [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) from another thread.</span></span> <span data-ttu-id="b49de-116">I stället måste fel förmedlas tillbaka till huvud tråden.</span><span class="sxs-lookup"><span data-stu-id="b49de-116">Instead, errors must be communicated back to the main thread.</span></span>

<span data-ttu-id="b49de-117">Det är möjligt att en cmdlet genererar ett undantag i dess implementering av metoden [system. Management. Automation. cmdlet. BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [system. Management. Automation.](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)cmdlet. ProcessRecord eller [system. Management. Automation. cmdlet. EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) .</span><span class="sxs-lookup"><span data-stu-id="b49de-117">It is possible for a cmdlet to throw an exception in its implementation of the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) method.</span></span> <span data-ttu-id="b49de-118">Alla undantag som har utlösts av dessa metoder (förutom några allvarliga fel villkor som stoppar Windows PowerShell-värden) tolkas som ett avslutande fel som stoppar pipelinen, men inte Windows PowerShell som helhet.</span><span class="sxs-lookup"><span data-stu-id="b49de-118">Any exception thrown from these methods (except for a few severe error conditions that stop the Windows PowerShell host) is interpreted as a terminating error which stops the pipeline, but not Windows PowerShell as a whole.</span></span> <span data-ttu-id="b49de-119">(Detta gäller endast för huvud-cmdlet-tråden.</span><span class="sxs-lookup"><span data-stu-id="b49de-119">(This applies only to the main cmdlet thread.</span></span> <span data-ttu-id="b49de-120">Ej fångade undantag i trådar som har skapats av cmdleten, i allmänhet, stoppar Windows PowerShell-värden.) Vi rekommenderar att du använder [system. Management. Automation. cmdlet. Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) i stället för att generera ett undantag eftersom fel posten innehåller ytterligare information om fel tillståndet, vilket är användbart för slutanvändaren.</span><span class="sxs-lookup"><span data-stu-id="b49de-120">Uncaught exceptions in threads spawned by the cmdlet, in general, halt the Windows PowerShell host.) We recommend that you use [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) rather than throwing an exception because the error record provides additional information about the error condition, which is useful to the end-user.</span></span> <span data-ttu-id="b49de-121">Cmdlets bör respektera den hanterade kod rikt linjen för att fånga och hantera alla undantag (`catch (Exception e)`).</span><span class="sxs-lookup"><span data-stu-id="b49de-121">Cmdlets should honor the managed code guideline against catching and handling all exceptions (`catch (Exception e)`).</span></span> <span data-ttu-id="b49de-122">Konvertera endast undantag av kända och förväntade typer till fel poster.</span><span class="sxs-lookup"><span data-stu-id="b49de-122">Convert only exceptions of known and expected types into error records.</span></span>

## <a name="see-also"></a><span data-ttu-id="b49de-123">Se även</span><span class="sxs-lookup"><span data-stu-id="b49de-123">See Also</span></span>

[<span data-ttu-id="b49de-124">System. Management. Automation. cmdlet. BeginProcessing</span><span class="sxs-lookup"><span data-stu-id="b49de-124">System.Management.Automation.Cmdlet.BeginProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)

[<span data-ttu-id="b49de-125">System. Management. Automation. cmdlet. EndProcessing</span><span class="sxs-lookup"><span data-stu-id="b49de-125">System.Management.Automation.Cmdlet.EndProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)

[<span data-ttu-id="b49de-126">System. Management. Automation. cmdlet. ProcessRecord</span><span class="sxs-lookup"><span data-stu-id="b49de-126">System.Management.Automation.Cmdlet.ProcessRecord</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)

[<span data-ttu-id="b49de-127">System. Management. Automation. PipelineStoppedException</span><span class="sxs-lookup"><span data-stu-id="b49de-127">System.Management.Automation.Pipelinestoppedexception</span></span>](/dotnet/api/System.Management.Automation.PipelineStoppedException)

[<span data-ttu-id="b49de-128">System. Management. Automation. cmdlet. Throwterminatingerror \*</span><span class="sxs-lookup"><span data-stu-id="b49de-128">System.Management.Automation.Cmdlet.Throwterminatingerror\*</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)

[<span data-ttu-id="b49de-129">System. Management. Automation. cmdlet. WriteError</span><span class="sxs-lookup"><span data-stu-id="b49de-129">System.Management.Automation.Cmdlet.WriteError</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)

[<span data-ttu-id="b49de-130">Windows PowerShell-felposter</span><span class="sxs-lookup"><span data-stu-id="b49de-130">Windows PowerShell Error Records</span></span>](./windows-powershell-error-records.md)

[<span data-ttu-id="b49de-131">Skriva en Windows PowerShell-cmdlet</span><span class="sxs-lookup"><span data-stu-id="b49de-131">Writing a Windows PowerShell Cmdlet</span></span>](./writing-a-windows-powershell-cmdlet.md)
